File: src/main/scala/io/github/configur8/Configuration.scala
package io.github.configur8

import io.github.configur8.Misconfiguration.throwUp

/**
 * A Configuration object is a successfully reified container for all of the properties that were defined.
 */
case class Configuration protected[configur8](private val propertySettings: Map[Property[_], String]) {

  /**
   * Public representation of the the settings
   */
  lazy val settings: Map[String, String] = propertySettings.map {
    case (p, v) => {
      (p.name, p.exposeMode.display(v))
    }
  }.toMap

  /**
   * Retrieve a property value which has been deserialized from it's string form
   * @param Property definition to retrieve
   * @tparam Type of the parameter
   * @return The deserialized parameter as an instance of the correct type
   */
  def valueOf[T](prop: Property[T]): T = prop.deserialize.apply(propertySettings.getOrElse(prop, throwUp("Unknown configuration key", prop.name).apply()))
}

File: java/src/main/java/io/github/configur8/Configuration.java
package io.github.configur8;

import java.util.Map;

import static java.util.stream.Collectors.toMap;

public class Configuration {

    private final Map<Property<?>, String> settings;

    Configuration(Map<Property<?>, String> settings) {
        this.settings = settings;
    }

    /**
     * Retrieve a property value which has been deserialized from it's string form
     * @param prop the property to retrieve the value for
     * @param <T> type
     * @return value of the property
     */
    public <T> T valueOf(Property<T> prop) {
        if (!settings.containsKey(prop)) {
            throw new Misconfiguration("Unknown configuration key '" + prop.name + "'");
        }
        return prop.deserialize.deserialize(settings.get(prop));
    }

    /**
     * Public representation of the the settings
     * @return settings described as a map with private values hidden
     */
    public Map<String, String> settings() {
        return settings.keySet().stream().collect(toMap(
                property -> property.name,
                k -> k.exposeMode.display(settings.get(k))
        ));
    }

}
